submit/routes/ui.py

```diff
@@ -0,0 +1,482 @@
+    if '.' in endpoint:
+        _, endpoint = endpoint.split('.', 1)
+    try:
+        stage = workflow.stage_from_endpoint(endpoint)
+    except ValueError:
+        stage = None
+
+    def get_current_stage_for_submission(submission: Submission) -> str:
+        """Get the endpoint of the current step for a submission."""
+        return get_workflow(submission).current_stage.endpoint
+
+    return {
+        'this_stage': stage,
+        'get_current_stage_for_submission': get_current_stage_for_submission
+    }
+
+
+@ui.context_processor
+def inject_workflow() -> Dict[str, Optional[workflow.Workflow]]:
+    """Inject the current workflow into the template rendering context."""
+    if hasattr(request, 'submission'):
+        return {'workflow': get_workflow(request.submission)}
+    return {'workflow': None, 'get_workflow': get_workflow}
+
+
+def add_immediate_alert(context: dict, severity: str,
+                        message: Union[str, dict], title: Optional[str] = None,
+                        dismissable: bool = True, safe: bool = False) -> None:
+    """Add an alert for immediate display."""
+    if safe and isinstance(message, str):
+        message = Markup(message)
+    data = {'message': message, 'title': title, 'dismissable': dismissable}
+
+    if 'immediate_alerts' not in context:
+        context['immediate_alerts'] = []
+    context['immediate_alerts'].append((severity, data))
+
+
+def handle(controller: Callable, template: str, title: str,
+           submission_id: Optional[int] = None,
+           get_params: bool = False, **kwargs) -> Response:
+    """
+    Generalized request handling pattern.
+
+    Parameters
+    ----------
+    controller : callable
+        A controller function with the signature ``(method: str, params:
+        MultiDict, session: Session, submission_id: int, token: str) ->
+        Tuple[dict, int, dict]``
+    template : str
+        HTML template to use in the response.
+    title : str
+        Page title, if not provided by controller.
+    submission_id : int or None
+    get_params : bool
+        If True, GET parameters will be passed to the controller on GET
+        requests. Default is False.
+    kwargs : kwargs
+        Passed as ``**kwargs`` to the controller.
+
+    Returns
+    -------
+    :class:`.Response`
+
+    """
+    logger.debug('Handle call to controller %s with template %s, title %s,'
+                 ' and ID %s', controller, template, title, submission_id)
+    if request.method == 'GET' and get_params:
+        request_data = MultiDict(request.args.items(multi=True))
+    else:
+        request_data = MultiDict(request.form.items(multi=True))
+
+    context = {'pagetitle': title}
+    try:
+        data, code, headers = controller(request.method, request_data,
+                                         request.session, submission_id,
+                                         **kwargs)
+    except (BadRequest, InternalServerError) as e:
```

Comment:
![Screenshot 2019-06-04 15 24 55](https://user-images.githubusercontent.com/746253/58908396-78db2400-86de-11e9-95eb-c1d8e54914d9.png)

I would suggest handling `BadRequest` separately and using a milder generic error message, since this isn't really unexpected and we probably don't want users contacting support in these cases.

Reply:
Most definitely. https://arxiv-org.atlassian.net/browse/ARXIVNG-2196